src/hooks/form/useSelectInstanceSpecs.ts

```diff
@@ -59,17 +61,94 @@ export function useSelectInstanceSpecs({
   nodeSpecs,
   ...rest
 }: UseSelectInstanceSpecsProps): UseSelectInstanceSpecsReturn {
-  const manager = useNodeManager()
+  const [state] = useAppState()
+  const {
+    manager: { nodeManager },
+  } = state
+
   const { defaultTiers } = useDefaultTiers({ type, gpuModel })
+  const { vouchers } = useAccountVouchers()
+
+  const filterValidNodeSpecs = useCallback(
+    (option: Tier) => {
+      if (paymentMethod === PaymentMethod.Hold) return option
+      if (!nodeSpecs) return false
+
+      return nodeManager.validateMinNodeSpecs(option, nodeSpecs)
+    },
+    [nodeManager, nodeSpecs, paymentMethod],
+  )
+
+  const disableHighTiersForHolding = useCallback(
+    (option: Tier) => {
+      if (paymentMethod !== PaymentMethod.Hold) return option
+      if (option.cpu <= 4) return option
+
+      return {
+        ...option,
+        disabled: true,
+        disabledReason: 'High tiers are only avaiable for Pay-as-you-go.',
+      }
+    },
+    [paymentMethod],
+  )
 
-  const options = useMemo(() => {
-    if (paymentMethod === PaymentMethod.Hold) return defaultTiers
-    if (!nodeSpecs) return []
+  const enableTiersWithVouchers = useCallback(
+    (option: Tier) => {
+      if (!option.disabled) return option
+      if (!vouchers.length) return option
+
+      console.log('vouchers', vouchers)
+      let enableOption = false
+
+      vouchers.some((voucher) => {
```

Comment:
Can you please build a dictionary filtering valid vouchers in a separated previous stage and cache it with useMemo for not doing the same calculations for each option